### PR TITLE
Propagate --test= for environments

### PR DESCRIPTION
**Pull request by haampie**
`--test={root,all}` was not propagated to environment installs, which seems like an oversight.

This PR allows you to run e.g.

```
spack -e path/to/env install --test=root --verbose
```

and it will run the tests of the environment root specs.

It also supports `spack -e xyz concretize --test=...`.

Maybe the environment should store a key `test: [root|all]` to ensure that concretization does not depend on a command line flag (e.g. running `spack -e myenv concretize` and then `spack -e myenv install --test=root` does not build test dependencies). But that issue can be fixed in a separate PR.

### CODE REVIEW

File: lib/spack/spack/cmd/concretize.py
Comment (becker33) [inline lib/spack/spack/cmd/concretize.py]: This isn't entirely accurate. It concretizes including test dependencies, but it doesn't ensure that the tests are actually run.

**Review by becker33 — CHANGES_REQUESTED** [review]
For the most part the code looks ok to me.

At a higher level, I know we had some discussions with @tldahlgren about revamping the test dependencies so they are left abstract at concretization time and concretized at test time.

@tldahlgren, is that something that you're pursuing as part of your ECP project, or is that fully on hold? I don't want to merge this if you're doing something that will solve the same problem in a very different (and more robust) way, but if that's on the back burner then I think we could get something like this in for the meantime.

**Comment by tldahlgren** [comment]
> For the most part the code looks ok to me.
> 
> At a higher level, I know we had some discussions with @tldahlgren about revamping the test dependencies so they are left abstract at concretization time and concretized at test time.
> 
> @tldahlgren, is that something that you're pursuing as part of your ECP project, or is that fully on hold? I don't want to merge this if you're doing something that will solve the same problem in a very different (and more robust) way, but if that's on the back burner then I think we could get something like this in for the meantime.

I'm afraid that effort is on the back burner.

File: lib/spack/spack/cmd/concretize.py
Comment (haampie) [reply lib/spack/spack/cmd/concretize.py]: Yeah, let me drop that, this is copy-pasta from the `install` command of course

**Comment by becker33** [comment]
@haampie this needs unit tests. Looking at the code it looks fine, so I don't expect the tests to catch anything, but it needs them anyway.

File: lib/spack/spack/cmd/install.py
Comment (becker33) [inline lib/spack/spack/cmd/install.py]: Can this segment be refactored  to be earlier in the method so it can be combined with the identical segment below?

File: lib/spack/spack/test/cmd/concretize.py
Comment (becker33) [inline lib/spack/spack/test/cmd/concretize.py]: These tests are redundant with the tests in env/cmd.py. Can you either

1. replace them with a test that just confirms the parser is setup properly for the concretize command? The rest of it is covered by the others.
2. extend them to parametrize by concretization separately/together and eliminate the tests in cmd/env.py

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/test/cmd/concretize.py
Comment (haampie) [reply lib/spack/spack/test/cmd/concretize.py]: I went for option 2

**Review by becker33 — APPROVED** [review]

**Comment by haampie** [comment]
Thanks! :)

**Pull request merged**
